### PR TITLE
Navigation: Allow a label to used for placeholder text

### PR DESCRIPTION
**Pull request by gwwar**
Changes here allow us to use a label as placeholder text instead of "Select Page". This makes creating navigation patterns more visually interesting and lets users more quickly pick out useful patterns. See related https://github.com/WordPress/gutenberg/pull/35063

https://user-images.githubusercontent.com/1270189/134546277-0b689eef-a62f-476d-956b-42ab5b01ceb3.mp4

### Testing Instructions

- Create a new post or page
- In the code editor add the following:
```
<!-- wp:navigation -->
<!-- wp:navigation-link {"label":"About","type":"page","kind":"post-type","isTopLevelLink":true} /-->
<!-- /wp:navigation -->
```
- Switch back to the Visual Editor
- Verify that newly added links still have the correct placeholder text like "Select page"
- Navigate to the Navigation Editor
- Add a Page Link
- Unlink
- Verify that all attributes have been removed (the same behavior as trunk).

### CODE REVIEW

**Comment by jasmussen** [comment]
Nice one, this works as intended:
Screenshot 2021-09-24 at 09 23 55

One question though, can we enable theme authors to create these patterns in the UI? There's already an "unlink" feature, which perhaps we could tweak so that the label remains? Right now unlinking removes both label and link:

nav

In making unlink just remove the link, arguably it'd be consistent with unlinking text:
unlink

**Comment by gwwar** [comment]
I've updated the unlink function to only remove id/url in https://github.com/WordPress/gutenberg/pull/35094/commits/61519ec20e63c8ced8c97fb27a144987360f8d50

When we publish, should a link with a label but no url be visible in the published view?

https://user-images.githubusercontent.com/1270189/134949796-05ac14e5-c48f-44b3-936c-43c2f6b29497.mp4

**Comment by jasmussen** [comment]
> When we publish, should a link with a label but no url be visible in the published view?

Great question. On the one hand my initial instinct was "no".

On the other hand, the hope is that the wavy underline/linting indicator of a dead link can scale to blocks like Buttons, where right now you _can_ publish a button that links no-where.

I would also like to pair the underline (separate effort) with a pre-publish panel that says "these blocks have dead links", making the wavy underline an extra affordance and not the sole indicator. In that light, maybe it's okay to publish? Not a strong opinion either way, and probably best to choose what feels right and go from there.

**Comment by gwwar** [comment]
@jasmussen If we follow up with a pre-publish check I think it's less of an issue to have this display with the URL empty. That's what current behavior does, but it's harder to get into this state since folks end up needing to edit the markup manually.

**Comment by getdave** [comment]
@gwwar I wonder if this could be used in conjunction with https://github.com/WordPress/gutenberg/pull/33849 so that if the user just enters the text it will use that as the `label` placeholder. Once they enter a `URL` then it updates fully.

**Comment by gwwar** [comment]
@getdave feel free to apply this changeset to the branch 👍  I was hoping that we could maybe land this piece first though to help enable pattern work.

**Comment by getdave** [comment]
Oh yes don't hold anything up. I just saw the potential synergy in the future 😄

**Comment by gwwar** [comment]
@jasmussen @getdave I think this piece is relatively safe to land by itself. Do y'all agree or is there any other behavior we should update in this PR?

**Review by jasmussen — APPROVED** [review]
Absolutely worth trying. Thank you!

**Comment by gwwar** [comment]
Looks like we need some input for what should happen in the Navigation Editor if we unlink a subset of properties for a navigation link.  In this branch it looks like we point the url at the menu item in that case.

https://user-images.githubusercontent.com/1270189/135515200-40ca9289-8bd0-4731-982a-e82c99d6cdbe.mp4

Note that serializing link placeholders in `trunk` has inconsistent behavior depending on the variation: https://github.com/WordPress/gutenberg/issues/35271

cc @getdave @tellthemachines @talldan any thoughts on ideal behavior for the Navigation Editor?

**Comment by getdave** [comment]
> cc @getdave @tellthemachines @talldan any thoughts on ideal behaviour for the Navigation Editor?

The Nav Editor should serialize the `url` to match whatever the URL is that was selected, not some generated slug of the Menu post type. I guess there is a bug... 🤔 

It looks like it's not possible to Unlink the nav item. If I unlink an item and the reload the Nav editor the unlinked item is restored as though no action occurred.

https://user-images.githubusercontent.com/444434/135609639-b5e62124-406c-4f23-88f2-ad2d12e91fa3.mov

Let's look at the block that we're persisting:

```js
{
    "clientId": "ea02ad66-b946-4c5f-8761-2aa7f7cd541c",
    "name": "core/navigation-link",
    "isValid": true,
    "attributes": {
        "label": "",
        "type": "",
        "id": "",
        "opensInNewTab": false,
        "url": "",
        "kind": "",
        "__internalRecordId": 807,
        "isTopLevelLink": true
    },
    "innerBlocks": []
}
```

and here is the block that comes back from the call to https://github.com/WordPress/gutenberg/blob/d079ba2c16495d6d976f5e08a7b5d259c3f13074/packages/edit-navigation/src/store/actions.js#L70-L72

```js
{
    "clientId": "ea02ad66-b946-4c5f-8761-2aa7f7cd541c",
    "name": "core/navigation-link",
    "isValid": true,
    "attributes": {
        "label": "my new draft page",
        "type": "page",
        "kind": "post-type",
        "url": "http://localhost:8888/my-new-draft-page/",
        "id": 770,
        "__internalRecordId": 807
    },
    "innerBlocks": []
}
```

It looks like perhaps the API ignores the "empty" block and just sends back the original so we can never remove the link.

@spacedmonkey Would you have any insight into this? Is the REST API rejecting menu items which don't have any attributes as invalid?

**Comment by spacedmonkey** [comment]
@getdave @gwwar Without looking at it deeply, I would guess it is these lines that might be the issue 
https://github.com/WordPress/gutenberg/blob/2bbb589b1c8449d0ddd3a922a42c5dc12eda2fbf/lib/class-wp-rest-menu-items-controller.php#L511-L519

Have you looked in the REST API response? If there a error returned?

It is worth noting, that menu items without a title should be removed, like the current menu editor. 



https://user-images.githubusercontent.com/237508/135726639-8e16edf4-c8b6-4c41-91c7-5bff018f1430.mov

**Comment by gwwar** [comment]
@getdave @spacedmonkey @jasmussen let's explore the Navigation Editor placeholder behavior in a separate PR to address https://github.com/WordPress/gutenberg/issues/35271

My current thoughts are that we ideally:

1. Prioritize retraining backwards compatibility for menu items
2. Provided that we can retain backwards compatibility, the most user friendly behavior I can think of adopting would be keeping the placeholders in the Navigation Editor view, but not render them on the published page.

Unless I'm misunderstanding the Navigation Editor project, there's hopefully some wiggle room to adopt new behaviors if it enhances the experience, instead of doing a straightforward functionality port.

I can keep this PR on hold until we resolve https://github.com/WordPress/gutenberg/issues/35271.

**Comment by gwwar** [comment]
🤔 to unblock this PR I can try to modify unlink behavior depending via a block parameter.

Edit: Reading through https://github.com/WordPress/gutenberg/issues/30007 it doesn't feel right to serialize an extra block attribute (via block context) to modify behavior that will only be seen in the edit view. Making each editor pass through an unlink function isn't ideal, either since it's easy to forget passing through the needed setting for a new editor, and the functions themselves will likely drift.

**Comment by gwwar** [comment]
Oh maybe if I use something like:

https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-navigation/src/filters/remove-edit-unsupported-features.js

**Comment by gwwar** [comment]
Okay I think I have this unblocked. In https://github.com/WordPress/gutenberg/pull/35094/commits/871418d79eda90551af066ed64904bb007b08662 I disable this new behavior for the navigation editor.

cc @getdave @jasmussen much appreciated if folks could take this for another spin

**Comment by jasmussen** [comment]
Here's a GIF showing the building of a nice header pattern, leveraging the "unlink" and "keep label" features in this branch:

nice

I think this is potent. I'd love for it to land. I'd defer to Dave on whether it needs an opt-out at all, but if that's what it takes to land this, it seems good to me since there's a clear comment in context, and the amount of code is manageable! 

Nice work as always.

**Review by getdave — COMMENTED** [review]
Sorry for the late response. GH comments becoming overwhelming.

I tested the editor and it is correctly disabled. 

https://user-images.githubusercontent.com/444434/137144015-4893b410-afee-4f84-a448-55197a456bf9.mp4

But I have questions about if this is the best way to proceed given the potential upcoming advances to the paradigm around how the Nav block and inner blocks interact.

Specifically I'm concerned about us starting to disable functionality in the editor via props on _inner_ nav blocks.

Why?

As you may have seen [the Nav block is possibly moving towards an architecture](https://wordpress.slack.com/archives/C01KDAZJMQ9/p1634131070324000) where the _presentation_ is governed by the wrapping `core/navigation` block whilst the inner blocks represent the _structure_ of the nav _data_. Think about the Nav block as being akin to a Template Part and how this facilitates reusable navigation data with different presentations.

The vision is that the Nav Editor becomes another instance of an Isolated Template Part editor, concerned only with editing the inner blocks. In this scenario there would be no wrapping `core/navigation` block used in the Nav Editor.

One of the benefits of this is that we'd be able to stop having to worry about features added to the Nav block leaking into the Editor and causing regressions. The inner blocks are "raw" so we shouldn't need to disable anything on them.

Therefore (and circling back to my concerns) it we start controlling the inner blocks via props then we're going to be back in the same situation of having to "undo" things on the inner blocks.

Is there a way this feature could be added via the `core/navigation` block itself - perhaps via some context mechanism. Ideally if the configuring attribute isn't provided to the parent Nav block then the feature shouldn't manifest on the inner block.

I'd like to get a 2nd opinion from @talldan on this one as may disagree. 

Edit: forgot to say I ❤️ the feature. Thank you for working on it. It will work very nicely with https://github.com/WordPress/gutenberg/pull/33849.

**Comment by gwwar** [comment]
> Is there a way this feature could be added via the core/navigation block itself - perhaps via some context mechanism. Ideally if the configuring attribute isn't provided to the parent Nav block then the feature shouldn't manifest on the inner block.

@getdave Thanks for taking another look! I did consider using block context, but this currently only works for block attributes. In this particular case, it didn't feel like the correct fit to serialize a new attribute on the navigation and navigation link, since this is modifying editor behavior only (and its primary use case is for dropping in easier to use patterns). 

Existing links without the attribute also won't behave as we expect for unsetting a link.

Overall, if we're a bit wary about diverging behavior, another option would be fixing behavior in https://github.com/WordPress/gutenberg/issues/35271 first. (Though it does sound like there are serialization issues to work through).

**Comment by jasmussen** [comment]
For a little bit of context, the TwentyTwentyTwo theme will come with some nice bundled header patterns, including a fresh version of [this one](https://user-images.githubusercontent.com/1204802/134635112-d22cb9e9-f8d3-4e8c-8d7f-0da2c29b8583.gif). That particular pattern is highly opinionated, and is designed around having precisely 4 menu items, no more, no less. And the pattern is that much more effective when the labels are pre-suggested, knowing a user can always rename them after wiring them up.

In that vein, I would really love for a version of this PR to be able to land for 5.9, so such patterns are possible. Given that horizon, is it better to disable it from the nav editor as is done now, or to fix it separately in the nav editor? What's the most feasible?

**Comment by gwwar** [comment]
> I would really love for a version of this PR to be able to land for 5.9, so such patterns are possible. Given that horizon, is it better to disable it from the nav editor as is done now, or to fix it separately in the nav editor? What's the most feasible?

Since there already needs to be a cleanup of the `remove-edit-unsupported-features` filter, adding one more prop doesn't seem too problematic for me, especially since folks are thinking through this problem. I am happy to defer to folks who have been working more closely on the navigation editor screens. Maybe cc @tellthemachines @talldan @draganescu

For context changes here add the ability to used named placeholders which is very helpful in creating interesting patterns, eg "Home, About, Contact" vs "add link, add link, add link". I'm proposing a fork in behavior here since the Editor Navigation screen doesn't use patterns yet and there are some unsolved inconsistencies with how placeholders are serialized in https://github.com/WordPress/gutenberg/issues/35271

**Comment by talldan** [comment]
The context I'm missing is why the behavior of 'Unlink' is being changed within the Navigation Block. Is it because that's the only way a pattern builder can build patterns? If so, I don't think that's right, the navigation block should be focusing on the primary use case for the majority of users, not pattern builders.

If it's being changed because everyone agrees that the right experience for all users would be for 'Unlink' to leave a label untouched, then it seems like a good change. But I think this is still debatable. For the link variations that are connected to a resource like a page, it might be quite frustrating for the label to remain present after 'Unlink', particularly if that label still matches the page title.

Then there's the case of the Navigation Editor, which I don't think should miss out on the best user experience because of past rules. Most of the features disabled in the navigation editor are like that because a theme can't possibly support those features. But this seems like a case that the navigation editor could support.

So to summarise, my opinion is that we need to understand how it should work for the majority of users, then implement that everywhere if we can.

**Comment by jasmussen** [comment]
To an extent, the navigation block is currently the odd one out. In the cases of every other link, unlinking the item just unlinks it, and does not clear any labels set:

unlink text

unlink buttons

In light of that, it feels odd that the navigation block does it:

navigation

The fact that it enables building patterns is a side-effect of adding consistency there. 

We have spent some time making it easy to build navigation people for most people, lately with the plus instantly inserting a page link, and the URL picker immediately opening to wire up a destination. Creating a menu is currently fairly seamless and smooth.

One of the benefits of the navigation block is that it lets you customize it after the fact: change labels if you like, for example prefixing "**New:**" to an item that's been updated, or to offer general customizability. You can even edit the URL after the fact, should you want to append a query string or fix a typo, or just point it somewhere else entirely. In light of this, the URL dialog is agnostic to the content, and it could be argued that it should work the same everywhere.

**Comment by talldan** [comment]
> To an extent, the navigation block is currently the odd one out. In the cases of every other link, unlinking the item just unlinks it, and does not clear any labels set

There's quite a strong difference in that those interfaces invite the user to type the label or paragraph text first and then add the link. In that way, linking and unlinking has symmetry.

That's not true in the Navigation Link block, the act of linking also creates the label in the majority of cases.

**Comment by jasmussen** [comment]
That [the symmetry of linking and unlinking] remains the case, though? This interface is only ever accessible after you've created the link in the first place, at which point you can change both the label and the link to break any symmetry, it feels like the unlink button is just an extension of that. Unlinking feels different to me than removing and reinserting a link block.

**Comment by talldan** [comment]
It is quite a minor thing, so I don't really feel that strongly about it. It just seems like a really unusual flow that the way to add an unlinked item is:
1. Add a link (which prefills the label)
2. Unlink the link
3. Modify the label that I didn't want in the first place.

What use case is this fulfilling for the majority of users?

Maybe this will get better with https://github.com/WordPress/gutenberg/issues/30170, but I am genuinely confused.

**Comment by jasmussen** [comment]
Being such an important part of a website, there's a great deal of value in being able to provide rich patterns as shortcuts to accomplishing good designs. In concert with <a href="https://github.com/WordPress/gutenberg/issues/32373">other features built for this purpose</a>, being able to create <a href="https://github.com/WordPress/gutenberg/issues/32373">an opinionated 4-item pattern</a> like this one can help enable or inspire rich designs. 

Screenshot 2021-10-15 at 09 58 19

Patterns, as noted, is a side effect. Being able to <a href="https://github.com/WordPress/gutenberg/issues/18866">unlink an item</a> on its own can be as important as being able to edit a link. Since you can already change labels and URLs after the fact, also keeping the behavior of the unlink button consistent serves to reinforce confidence in the UI.

Flipping the question on its head: what use case is fulfilled by keeping the behavior of the unlink button unique to the menu item?

**Comment by talldan** [comment]
I think the main part I'm concerned about is that 'Unlink' potentially now introduces a system that can be abused by users to add semantically incorrect tags to their websites (anchors without hrefs being used as plain text), something that was previously quite strongly discouraged in the block by the use of the `Add link` button.

A couple of thoughts:
- Is this definitely needed? Patterns can use valid links, but they might be a document anchor link (`#`).
- What are the alternatives? How about a `placeholder` prop, or an `isPlaceholder` state for the block? The placeholder could be something that is only shown when the block is unselected and is unset when the user adds their own link.

**Comment by talldan** [comment]
> Being able to unlink an item on its own can be as important as being able to edit a link.

Why? The block itself is called 'link', it seems unusual to consider it valid in an unlinked state.

For the issue you linked to, the reporter had a very clear use case, they wanted to create a submenu where the top-level item isn't a link. This has since been solved by the submenu block, which keeps the semantics of the link block in-tact.

**Comment by jasmussen** [comment]
Thanks for talking through this with me. As outlined, I believe there's value to be found, but given the concerns voiced, perhaps we should pause this one after all. Of note, the following is possible as a pattern today:

Screenshot 2021-10-14 at 13 43 20

Given that fact gets us 90% there, it could be a good baseline to start from and get some feedback on patterns as they are being created and used. We can then revisit in the future if we find a need to take it further.

**Comment by mtias** [comment]
Chiming in a bit late, but I don't think we should ever allow a state like this:

image

Where it's not clear at all if things are in "placeholder" state or there is actual content set. If I have pages that coincide with those names, I might mistake the above as being set already while that's not true. Placeholder states that need the user to interact should be obvious and uncomplicated.

**Comment by jasmussen** [comment]
Understood, I might have stared at this for too long. Thanks all, for chiming in and for exploring this one, I'll go ahead and close the adjacent issues.

**Comment by mtias** [comment]
I think our current solution is not quite right, so we'll need to continue to stare at it but at larger intervals :)

**Pull request closed without merging**
